### PR TITLE
Add blog article: Howto migrate a real-life batch pipeline from the DataSet API to the DataStream API

### PR DESCRIPTION
**Pull request by echauchot**
Cross posting to official flink blog from https://echauchot.blogspot.com/2022/11/flink-howto-migrate-real-life-batch.html. This blog article was already submitted to flink community for validation before publishing to my personal blog.

R: @zentol

### CODE REVIEW

File: content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html
Comment (zentol) [inline content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html]: Please remove the .html files from the PR. Let the committer generate these once the merge is happening.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]:
```suggestion
Instead of going through all of the code which is
available [here](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java)
 we will rather focus on some key areas of the migration. The code is based on the latest release
```

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]:
```suggestion
Now that we have a working triggering, we need to call a standard join with the  _Row::join_
```

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: > unless it will not materialize the collection of data

Not sure what this is referring to.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]:
```suggestion
Another thing: to be able to use Flink state, we need to key the datastream beforehand, even if there
```

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: This seems weird. It's super sketchy for a function to just close() itself.

Why not map records to "1" and apply a reduce function that maxes out at 100?

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: Let's point to a specific commit instead of the master branch

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: Why aren't you linking to the Flink javadocs at https://nightlies.apache.org/flink/flink-docs-release-1.16/api/java/?

**Review by zentol — COMMENTED** [review]
hmmm.......my impression is that some of things described here feel...sketchy? I don't know the batch execution mode enough for whether this is just the way things have to be done.

@alpinegizmo What do you think?

File: content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html
Comment (echauchot) [reply content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html]: Yes I know. For some reasons I messed up my last commit's `git add`. That being said I understand why `content` directory is not in the git ignore, it is because you commit it as part of the website rebuild before merging right ?

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: because at the time I wrote this article I did not know that javadocs were also in nightlies and I forgot to change the links in this old article. Thanks for catching it, I'll update all the links in this article.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: yes, I did that for all the other links in the other articles, thanks for catching this missing one !

File: content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html
Comment (echauchot) [reply content/2023/04/12/howto-migrate-a-real-life-batch-pipeline-from-the-dataset-api-to-the-datastream-api/index.html]: done

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: done

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: done

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: Wording issue. I meant "except  it will not materialize the collection of data".
Correct me if I'm wrong but DataSet#groupBy() shuffles the data per key and materializes the data belonging to the same key in memory (in a collection) whereas DataStream#keyBy() doesn't do this materialization.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: I agree it is sketchy. I did that to be sure to keep the order built in the previous step. That being said, I could replace by a `ProcessFunction` that does the limit. Indeed, I tested : there is no rebalancing of data (just a hash) between the order by step and the limit step in Flink execution graph.  

Also please note that I could have done much simpler code by mixing DataStream code with SQL/Table API especially using `LIMIT 100` and `ORDER BY` but the point was to try to code the equivalent of DataSet features with pure DataStream API to prepare the migration path for users (discover limitations such as [this one](https://issues.apache.org/jira/browse/FLINK-22587)) and also to compare the performance of both APIs.

**Comment by echauchot** [comment]
@zentol I have fixed a bug I mentioned earlier : missing windowing (it was not obvious to me that DataStream API in batch mode required windowing even if there is no event timestamps in the data, but nevermind). I also addressed your comments. 
PTAL.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: > DataSet#groupBy() [...] materializes the data belonging to the same key in memory

That depends on what you're doing afterwards.
A reduce doesn't materialize everything and works similarly to a streaming reduce.

For a join, one/both sides are materialized in memory and/or spilled to disk.
Group Reduce materializes all data.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: we shouldnt need an alias; this is only required for older blog posts.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (zentol) [inline docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]:
```suggestion
is [this](https://github.com/echauchot/tpcds-benchmark-flink/blob/f342c1983ec340e52608eb1835e85c82c8ece1d2/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDataset.java)
batch pipeline that implements the above query using the DataSet API
```

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: Makes sense. So in my case, as the following operator is a reduce, there will be indeed no materialization. Thanks for the details.

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: fixed

**Comment by echauchot** [comment]
@zentol finished updating PTAL

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: the details were added to the article, did you see it ?

File: docs/content/posts/2023-04-12-howto-migrate-to-datastream.md
Comment (echauchot) [reply docs/content/posts/2023-04-12-howto-migrate-to-datastream.md]: Limit code was remade and article updated. Did you see it ?

**Comment by echauchot** [comment]
@zentol how does this article look like after last modifications ?

**Review by zentol — APPROVED** [review]

**Pull request merged**
